src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,321 @@
+ * and redistribute Covered Software under those terms instead of these.   *
+ *                                                                         *
+ * Source is provided to this software because we believe users have a     *
+ * right to know exactly what a program is going to do before they run it. *
+ * This also allows you to audit the software for security holes.          *
+ *                                                                         *
+ * Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+ * and add new features.  You are highly encouraged to submit your changes *
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <config.h>
+#include <filesystem>
+#include <array>
+#include <sstream>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <libvmi/libvmi.h>
+
+#include "ipt.h"
+#include "plugins/output_format.h"
+#include "private.h"
+
+namespace
+{
+
+uint64_t pack_payload(uint32_t cmd, uint32_t data)
+{
+    return (static_cast<uint64_t>(cmd) << 32) | data;
+}
+
+void emit_ptwrite64(std::ofstream& stream, uint64_t payload)
+{
+    // ptwrite packet
+    stream.put(0x02);
+    // no FUP, 8 bytes of payload
+    stream.put(0x32);
+    stream.write(reinterpret_cast<char*>(&payload), sizeof(payload));
+}
+
+
+event_response_t ipt_cr3_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<ipt>(info);
+    auto& vcpu = plugin->vcpus[info->vcpu];
+
+    vcpu.flush(info->regs->vmtrace_pos);
+
+    vcpu.annotate(pack_payload(PTW_CURRENT_CR3, info->regs->cr3));
+    vcpu.annotate(pack_payload(PTW_CURRENT_TID, info->proc_data.tid));
```

Comment:
Do we not annotate PID?

Reply:
Generally, PID can be deduced from CR3/TID